src/pipeformer/internal/structures.py

```diff
@@ -0,0 +1,298 @@
+from typing import Dict, Iterable, Optional, Set
+
+import attr
+import oyaml as yaml
+from attr.validators import deep_iterable, deep_mapping, instance_of, optional
+from troposphere import Ref, Sub, Template, cloudformation, secretsmanager, ssm
+
+from .util import reference_name, resource_name
+
+__all__ = ("Config", "PipelineStage", "PipelineAction", "Input", "ProjectTemplates", "WaitConditionStack", "Pipeline")
+_STRING_STRING_MAP = deep_mapping(key_validator=instance_of(str), value_validator=instance_of(str))
+
+
+def _resolve_parameter(name: Ref, version: str) -> Sub:
+    """Build a CloudFormation dynamic reference string structure that resolves a SSM Parameter.
+
+    :param name: Parameter name
+    :param version: Parameter version
+    :return: Dynamic reference
+    """
+    return Sub(f"{{{{resolve:ssm:${{name}}:{version}}}}}", {"name": name})
+
+
+def _resolve_secret(arn: Ref) -> Sub:
+    """Build a CloudFormation dynamic reference string structure that resolves a Secrets Manager secret.
+
+    :param arn: Secret Arn
+    :return: Dynamic reference
+    """
+    return Sub(f"{{{{resolve:secretsmanager:${{arn}}:SecretString}}}}", {"arn": arn})
+
+
+@attr.s
+class WaitConditionStack:
+    """Container to hold all resources for a wait-condition-initiated stack.
+
+    :param condition: Wait condition
+    :param handle: Wait condition handle
+    :param stack: Stack
+    """
+
+    condition: cloudformation.WaitCondition = attr.ib(validator=instance_of(cloudformation.WaitCondition))
+    handle: cloudformation.WaitConditionHandle = attr.ib(validator=instance_of(cloudformation.WaitConditionHandle))
+    stack: cloudformation.Stack = attr.ib(validator=instance_of(cloudformation.Stack))
+
+
+class _ConfigStructure:
+    """Base for configuration structures."""
+
+    @staticmethod
+    def _clean_kwargs(kwargs: Dict):
+        """Convert keys separators from YAML-valid "-" characters to Python-variable-name-valid "_" characters."""
+        return {key.replace("-", "_"): value for key, value in kwargs.items()}
+
+    @classmethod
+    def from_dict(cls, kwargs: Dict):
+        """Load from a dictionary."""
+        return cls(**cls._clean_kwargs(kwargs))
+
+
+@attr.s
+class Input(_ConfigStructure):
+    """Container and formatter for input values.
+
+    :param name: Input name
+    :param description: Input description
+    :param secret: Is this input a secret?
+    :param value: Input value (optional)
+    """
+
+    name: str = attr.ib(validator=instance_of(str))
+    description: str = attr.ib(validator=instance_of(str))
+    secret: bool = attr.ib(validator=instance_of(bool))
+    value: Optional[str] = attr.ib(default=None, validator=optional(instance_of(str)))
+    version = None
+
+    def __attrs_post_init__(self):
+        """Set additional configuration values based on input type."""
+        if self.secret:
+            self._resource_type = secretsmanager.Secret
+            self._value_type = "Arn"
```

Comment:
Is something forcing these to be stringly typed?

Reply:
Added a typehint for this class-level variable.